add PATCH request on /metadata now supports multiple values

minurmin opened the pull request:
## References

* Related to DSpace/dspace-angular#2056 (and fixes it if [DSpace/dspace-angular/#3078](https://github.com/DSpace/dspace-angular/pull/3078) was implemented with PATCH add instead of PATCH replace)
* Related to DSpace/RestContract#304

## Description

PATCH add to /metadata currently supports adding only one value at time. If values are already present new value is appended only to the start or end of value list (not between, even if /metadata/field.name/n index is given.

In addition, if multiple values are given as a list (to initialize a metadata field without a given index), only one value of the list is picked, conflicting [RFC section 4.1](https://tools.ietf.org/html/rfc6902#section-4.1) and REST contract. This PR - partially based on #9610 by @MMilosz (applied to PATCH replace) - should bring the behaviour in line with the REST contract, providing additional error checks.

## Instructions for Reviewers

List of changes in this PR:
* New method extractMetadataValueListFromOperation has been added to helper class  DSpaceObjectMetadataPatchUtils.java. This is a generalization to earlier method extractMetadataValueFromOperation (used by both add and replace operations) that can handle both individual values or arrays.
* DSpaceObjectMetadataAddOperation.java now handles lists (with one-value-list as special case to handle individual values). Multivaluelist is accepted only if no index is given (i.e. you can add a list with `/metadata/dc.title` but not with `/metadata/dc.title/0`), otherwise this is considered a 400 Bad request.
* Logic and related to metadata additions with indexes are improved. Earlier, unless `-1` value was given, new metadata value was always added to the start of the list (i.e. index `0` - even if PATCH called given with another index number parseInt was not present). Now the new value is appended to position w.r.t. given index. For example, if two different dc.titles are present, call `/metadata/dc.title/1` would append the new value between the first and last values.
* The following checks to handling index values have been added: index number should not be greater that the total number of existing metadata values and it should be a number. I.e. call  `/metadata/dc.title/20` would result to 422 Unprocessable entity and call  `/metadata/dc.title/asdf` would result to 400 Bad request.

This PR can be tested with any entity that uses metadata. For example, updating metadata to an existing community (as in #9610 but applied to add operation) with curl. 

`$ curl http://localhost:8080/server/api/core/communities/:UUID -X PATCH --data-raw '[{"op":add", ... }
`

Simple example tests to try with https://github.com/the-library-code/dspace-rest-python :

```python
d = DSpaceClient(api_endpoint=URL, username=USERNAME, password=PASSWORD, fake_user_agent=False)

#...

dso = d.resolve_identifier_to_dso(handle)

# error 400 - multiple values with place
d.api_patch(url=dso.links["self"]["href"],operation=DSpaceClient.PatchOperation.ADD, path='/metadata/dc.description.abstract/0', value= [ { "value": "Kuvaus suomeksi", "language": "fi" }, { "value": "Description in English", "language": "en" }])

#success - single value in list
d.api_patch(url=dso.links["self"]["href"],operation=DSpaceClient.PatchOperation.ADD, path='/metadata/dc.description.abstract/0', value= [ { "value": "single value"} ])

#original dc.description.abstract values are replaced
d.api_patch(url=dso.links["self"]["href"],operation=DSpaceClient.PatchOperation.ADD, path='/metadata/dc.description.abstract', value= [ { "value": "Kuvaus suomeksi", "language": "fi" }, { "value": "Description in English", "language": "en" }])

#replace existing values as one (as no place is given)
d.add_metadata(dso,'dc.description.abstract','replaced',language='en')

#single value to start of list
d.add_metadata(dso,'dc.description.abstract','Description in English',language='en',place='0')

#single value to end of list
d.add_metadata(dso,'dc.description.abstract','Kuvaus suomeksi',language='fi',place='-1')

#single value in between
d.add_metadata(dso,'dc.description.abstract','intermediate value',place='1')

#results to error 422, out of bounds
d.add_metadata(dso,'dc.description.abstract','erroneous place',language='fi',place='10')

#results to error 400, invalid place
d.add_metadata(dso,'dc.description.abstract','erroneous place',language='fi',place='adsf')

#fails (metadatafield does not exist)
d.add_metadata(dso,'dc.newfield','uninitialized new field')
```

## Checklist

- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [x] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR **passes Checkstyle** validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR **includes Javadoc** for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR **passes all tests and includes new/updated Unit or Integration Tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

MMilosz committed:
fix: PATCH request on /metadata now supports multiple values

Original change related to PR #9610 applied only to DSpaceObjectMetadataPatchUtils, not replace operation.

minurmin committed:
PATCH add now supports adding/replacing multiple values with additional index checks

To better comply with RFC 6902.
If multiple values (without index) are given, these are added (and earlier values replaced, if any) to given metadata field.
When a single value is given, index is checked (number format and index bounds) and value is appended to position based on index (not only 0 or -1).
It is now considered an error if both index and list of multiple values are given.
Accommodates changes in DSpace 10.

MMilosz commented:
@minurmin thanks a lot for supporting my changes from #9610 when it was active and for cherry-picking my commit into this PR–really appreciate it!

I think we're on the same page regarding RFC Section 4.3–that "an indivual value" could reasonably be interpreted as either _one object_ OR _one array_. For me, that nuance is important

Also, I completely agree that improving the `add` operation is a great idea–my only concern is that this change might surface issues in the frontend. Particularly, I believe there might be places where it's assumed that `add` always accepts the first element when an array is passed, and it's probably worth investigating further to make sure nothing breaks, unexpectedly

minurmin committed:
Add test data to demonstrate adding multiple values at once and adding metadatavalue between others

Executed e.g. as part of org.dspace.app.rest.ItemRestRepositoryIT.

lgeggleston commented:
@MMilosz @Micheleboychuk General reminder for all new feature PRs currently under review: there is still a possibility of getting this into 10.0, in case you have the capacity to complete the review at the moment.   Feature merge deadline will be Friday 3/27. Thank you!

Micheleboychuk commented on dspace-server-webapp/src/test/resources/org/dspace/app/rest/test/metadata-patch-suite.json:
```diff
@@ -138,6 +138,44 @@
         ]
       }
     },
+    {
```
Thanks for adding the new test cases to the suite — they cover the main scenarios well! There are a few additional cases that could help ensure full coverage of the new behavior:   
  
 - add with a single value (not an array) and no index, when values already exist — just to make explicit what the expected behavior is in this case
 - add with a numeric index equal to the current size of the list (e.g. /dc.title/2 when 2 values exist), to verify appending at the end via explicit position works correctly

minurmin replied:
Added respective test cases to metadata-patch-suite.json.

minurmin commented:
@Micheleboychuk thank you for the review! Unfortunately I haven't got time to revise the PR yet, but your suggestions seem sound and I shall incorporate them to the next version. Because of the recent merge conflict some additional restructuring may be needed as well.

minurmin committed:
Address feedback from code review

 - Removed unused method.
 - Improved error message.
 - Added check for negative index numbers with simplified index check logic.
 - Added additional test cases.

minurmin commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/patch/operation/DSpaceObjectMetadataAddOperation.java:
```diff
@@ -60,15 +65,58 @@ public R perform(Context context, R resource, Operation operation) throws SQLExc
      */
     private void add(Context context, DSpaceObject dso, DSpaceObjectService dsoService, MetadataField metadataField,
```
Deleted in newest commit.

minurmin commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/patch/operation/DSpaceObjectMetadataAddOperation.java:
```diff
@@ -60,15 +65,58 @@ public R perform(Context context, R resource, Operation operation) throws SQLExc
+                    metadataField.getMetadataSchema().getName(), metadataField.getElement(),
+                    metadataField.getQualifier(), Item.ANY);
+            if (indexInt > metadataValues.size()) {
+                throw new UnprocessableEntityException("There is no metadata of this type at that index");
```
Rephrased error message.

minurmin commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/patch/operation/DSpaceObjectMetadataAddOperation.java:
```diff
@@ -60,15 +65,58 @@ public R perform(Context context, R resource, Operation operation) throws SQLExc
             indexInt = -1;
+        } else if (index != null) {
+            try {
+                indexInt = Integer.parseInt(index);
```
Negative check added with additional streamlining to if structure (`indexInt -1` is now used as default, eliminating one if block).

minurmin commented:
@Micheleboychuk The PR has been rebased to main with additions based on your review. I interpreted RFC 6902 for the test case *add with a single value (not an array) and no index, when values already exist* such that existing value(s) are replaced (i.e. as if an array with single value was given).
